File: ros/catkin_ws/src/simple_robotic_arm/scripts/opencv_obstacle_avoidance.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Lightweight monocular obstacle avoidance for ROS 1 + OpenCV.

This is a heuristic, no-training node. It assumes the bottom strip of the image
mostly contains floor, then treats pixels that differ strongly from that local
floor model inside a forward ROI as possible obstacles. The ROI is split into
left/center/right sectors and the node publishes geometry_msgs/Twist:
  - center blocked: stop forward motion and turn toward the clearer side
  - side blocked only: keep moving slowly and bias away from the obstacle
  - clear: move forward

Limitations: monocular RGB does not provide reliable metric distance. For
safety-critical use, pair this with depth/LiDAR/bump sensors and a hardware E-stop.
"""
from __future__ import print_function

import sys

import cv2
import numpy as np
import rospy
from cv_bridge import CvBridge, CvBridgeError
from geometry_msgs.msg import Twist
from sensor_msgs.msg import Image


def clamp(value, low, high):
    return max(low, min(high, value))


class OpenCVObstacleAvoidance(object):
    def __init__(self):
        self.bridge = CvBridge()
        self.image_topic = rospy.get_param("~image_topic", "/usb_cam/image_raw")
        self.cmd_topic = rospy.get_param("~cmd_vel_topic", "/cmd_vel")
        self.debug_topic = rospy.get_param("~debug_image_topic", "~debug")
        self.publish_debug = bool(rospy.get_param("~publish_debug_image", False))

        # Motion. Keep defaults conservative because this is monocular vision only.
        self.forward_speed = float(rospy.get_param("~forward_speed", 0.08))
        self.slow_speed = float(rospy.get_param("~slow_speed", 0.04))
        self.turn_speed = float(rospy.get_param("~turn_speed", 0.45))
        self.side_steer_gain = float(rospy.get_param("~side_steer_gain", 0.35))
        self.stop_on_no_image = bool(rospy.get_param("~stop_on_no_image", True))
        self.no_image_timeout = float(rospy.get_param("~no_image_timeout", 1.0))

        # Regions are normalized image coordinates.
        self.floor_sample_y0 = float(rospy.get_param("~floor_sample_y0", 0.82))
        self.floor_sample_y1 = float(rospy.get_param("~floor_sample_y1", 0.96))
        self.roi_y0 = float(rospy.get_param("~roi_y0", 0.38))
        self.roi_y1 = float(rospy.get_param("~roi_y1", 0.88))
        self.roi_x0 = float(rospy.get_param("~roi_x0", 0.12))
        self.roi_x1 = float(rospy.get_param("~roi_x1", 0.88))

        # Mask thresholds. Tune these per camera/floor lighting.
        self.hsv_floor_dist_thresh = float(rospy.get_param("~hsv_floor_dist_thresh", 42.0))
        self.edge_thresh = int(rospy.get_param("~edge_thresh", 70))
        self.dark_v_thresh = int(rospy.get_param("~dark_v_thresh", 55))
        self.min_saturation = int(rospy.get_param("~min_saturation", 35))
        self.open_kernel = int(rospy.get_param("~open_kernel", 3))
        self.close_kernel = int(rospy.get_param("~close_kernel", 7))

        # Decision thresholds are ratios of obstacle pixels in each sector.
        self.center_stop_ratio = float(rospy.get_param("~center_stop_ratio", 0.18))
        self.side_slow_ratio = float(rospy.get_param("~side_slow_ratio", 0.24))
        self.clear_hysteresis = float(rospy.get_param("~clear_hysteresis", 0.04))
        self.min_obstacle_area = int(rospy.get_param("~min_obstacle_area", 600))

        self.last_image_time = rospy.Time(0)
        self.last_blocked = False

        self.twist_pub = rospy.Publisher(self.cmd_topic, Twist, queue_size=1)
        self.debug_pub = None
        if self.publish_debug:
            self.debug_pub = rospy.Publisher(self.debug_topic, Image, queue_size=1)

        self.sub = rospy.Subscriber(self.image_topic, Image, self.on_image, queue_size=1, buff_size=2 ** 24)
        self.timer = rospy.Timer(rospy.Duration(0.1), self.on_timer)
        rospy.on_shutdown(self.publish_stop)
        rospy.loginfo("opencv_obstacle_avoidance: image=%s cmd=%s debug=%s", self.image_topic, self.cmd_topic, self.publish_debug)

    def publish_stop(self):
        self.twist_pub.publish(Twist())

    def on_timer(self, _event):
        if not self.stop_on_no_image:
            return
        if self.last_image_time == rospy.Time(0):
            return
        age = (rospy.Time.now() - self.last_image_time).to_sec()
        if age > self.no_image_timeout:
            rospy.logwarn_throttle(2.0, "obstacle_avoidance: no image for %.2fs, stopping", age)
            self.publish_stop()

    def clip_box(self, w, h):
        x0 = int(clamp(self.roi_x0, 0.0, 1.0) * w)
        x1 = int(clamp(self.roi_x1, 0.0, 1.0) * w)
        y0 = int(clamp(self.roi_y0, 0.0, 1.0) * h)
        y1 = int(clamp(self.roi_y1, 0.0, 1.0) * h)
        if x1 <= x0:
            x0, x1 = 0, w
        if y1 <= y0:
            y0, y1 = int(0.4 * h), h
        return x0, y0, x1, y1

    def floor_model(self, hsv):
        h, w = hsv.shape[:2]
        y0 = int(clamp(self.floor_sample_y0, 0.0, 1.0) * h)
        y1 = int(clamp(self.floor_sample_y1, 0.0, 1.0) * h)
        x0 = int(0.25 * w)
        x1 = int(0.75 * w)
        sample = hsv[y0:y1, x0:x1]
        if sample.size == 0:
            sample = hsv[int(0.8 * h):h, :]
        return np.median(sample.reshape(-1, 3), axis=0).astype(np.float32)

    def obstacle_mask(self, bgr):
        hsv = cv2.cvtColor(bgr, cv2.COLOR_BGR2HSV)
        gray = cv2.cvtColor(bgr, cv2.COLOR_BGR2GRAY)
        floor = self.floor_model(hsv)

        # Hue wraps at 180 in OpenCV HSV, so use circular hue distance.
        hsv_f = hsv.astype(np.float32)
        dh = np.abs(hsv_f[:, :, 0] - floor[0])
        dh = np.minimum(dh, 180.0 - dh) * 1.4
        ds = np.abs(hsv_f[:, :, 1] - floor[1]) * 0.35
        dv = np.abs(hsv_f[:, :, 2] - floor[2]) * 0.55
        color_dist = dh + ds + dv

        edges = cv2.Canny(gray, self.edge_thresh, self.edge_thresh * 2)
        non_floor = color_dist > self.hsv_floor_dist_thresh
        dark = hsv[:, :, 2] < self.dark_v_thresh
        saturated = hsv[:, :, 1] > self.min_saturation
        mask = (non_floor & (saturated | dark | (edges > 0))).astype(np.uint8) * 255

        if self.open_kernel >= 3:
            k = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (self.open_kernel, self.open_kernel))
            mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, k)
        if self.close_kernel >= 3:
            k = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (self.close_kernel, self.close_kernel))
            mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, k)
        return mask

    def sector_ratios(self, mask, box):
        x0, y0, x1, y1 = box
        roi = mask[y0:y1, x0:x1]
        if roi.size == 0:
            return 0.0, 0.0, 0.0, roi

        # Remove tiny speckles before measuring sector occupancy.
        contours, _ = cv2.findContours(roi, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        cleaned = np.zeros_like(roi)
        for c in contours:
            if cv2.contourArea(c) >= self.min_obstacle_area:
                cv2.drawContours(cleaned, [c], -1, 255, thickness=cv2.FILLED)

        rw = cleaned.shape[1]
        a = rw // 3
        left = cleaned[:, :a]
        center = cleaned[:, a:2 * a]
        right = cleaned[:, 2 * a:]

        def ratio(part):
            return float(np.count_nonzero(part)) / float(part.size) if part.size else 0.0

        return ratio(left), ratio(center), ratio(right), cleaned

    def decide(self, left, center, right):
        stop_threshold = self.center_stop_ratio
        if self.last_blocked:
            stop_threshold = max(0.01, stop_threshold - self.clear_hysteresis)

        twist = Twist()
        blocked = center >= stop_threshold
        if blocked:
            twist.linear.x = 0.0
            twist.angular.z = self.turn_speed if left < right else -self.turn_speed
            state = "blocked_turn_left" if left < right else "blocked_turn_right"
        elif left >= self.side_slow_ratio or right >= self.side_slow_ratio:
            twist.linear.x = self.slow_speed
            steer = clamp((left - right) * self.side_steer_gain, -self.turn_speed, self.turn_speed)
            twist.angular.z = -steer
            state = "side_avoid"
        else:
            twist.linear.x = self.forward_speed
            twist.angular.z = 0.0
            state = "clear"

        self.last_blocked = blocked
        return twist, state

    def on_image(self, msg):
        self.last_image_time = rospy.Time.now()
        try:
            bgr = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
        except CvBridgeError as exc:
            rospy.logwarn("obstacle_avoidance cv_bridge: %s", exc)
            return

        h, w = bgr.shape[:2]
        box = self.clip_box(w, h)
        mask = self.obstacle_mask(bgr)
        left, center, right, cleaned_roi = self.sector_ratios(mask, box)
        twist, state = self.decide(left, center, right)
        self.twist_pub.publish(twist)

        rospy.loginfo_throttle(
            1.0,
            "obstacle_avoidance: %s left=%.2f center=%.2f right=%.2f v=%.2f wz=%.2f",
            state,
            left,
            center,
            right,
            twist.linear.x,
            twist.angular.z,
        )

        if self.publish_debug and self.debug_pub is not None:
            self.publish_debug_image(msg, bgr, box, cleaned_roi, left, center, right, state)

    def publish_debug_image(self, msg, bgr, box, cleaned_roi, left, center, right, state):
        dbg = bgr.copy()
        x0, y0, x1, y1 = box
        overlay = dbg[y0:y1, x0:x1].copy()
        red = np.zeros_like(overlay)
        red[:, :, 2] = cleaned_roi
        overlay = cv2.addWeighted(overlay, 0.65, red, 0.35, 0.0)
        dbg[y0:y1, x0:x1] = overlay
        cv2.rectangle(dbg, (x0, y0), (x1 - 1, y1 - 1), (0, 255, 255), 2)
        sx = (x1 - x0) // 3
        cv2.line(dbg, (x0 + sx, y0), (x0 + sx, y1), (255, 255, 0), 1)
        cv2.line(dbg, (x0 + 2 * sx, y0), (x0 + 2 * sx, y1), (255, 255, 0), 1)
        label = "{} L:{:.2f} C:{:.2f} R:{:.2f}".format(state, left, center, right)
        cv2.putText(dbg, label, (10, 28), cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2, cv2.LINE_AA)
        try:
            out = self.bridge.cv2_to_imgmsg(dbg, encoding="bgr8")
            out.header = msg.header
            self.debug_pub.publish(out)
        except CvBridgeError as exc:
            rospy.logwarn("obstacle_avoidance debug cv_bridge: %s", exc)


def main():
    rospy.init_node("opencv_obstacle_avoidance", anonymous=False)
    try:
        OpenCVObstacleAvoidance()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: MasterComputer/ros/catkin_ws/src/simple_robotic_arm/src/opencv_obstacle_avoidance.cpp
/*
|--------------------------------------------------------------------------
| OpenCV 单目避障节点
|--------------------------------------------------------------------------
| 使用地面颜色采样、边缘和暗区启发式识别前方障碍物，并发布保守速度指令。
|--------------------------------------------------------------------------
*/

#include <algorithm>
#include <array>
#include <string>
#include <vector>

#include <cv_bridge/cv_bridge.h>
#include <geometry_msgs/Twist.h>
#include <opencv2/imgproc.hpp>
#include <ros/ros.h>
#include <sensor_msgs/Image.h>

namespace simple_robotic_arm {

double clampValue(double value, double low, double high) {
  return std::max(low, std::min(high, value));
}

/*
|--------------------------------------------------------------------------
| 避障决策节点
|--------------------------------------------------------------------------
| 把前方 ROI 分为左中右三块，中心阻挡则停下转向，侧边阻挡则慢速偏航避开。
|--------------------------------------------------------------------------
*/
class OpenCVObstacleAvoidanceNode {
 public:
  OpenCVObstacleAvoidanceNode() : private_nh_("~") {
    loadParams();
    twist_pub_ = nh_.advertise<geometry_msgs::Twist>(cmd_topic_, 1);
    if (publish_debug_) {
      debug_pub_ = nh_.advertise<sensor_msgs::Image>(debug_topic_, 1);
    }
    image_sub_ = nh_.subscribe(image_topic_, 1, &OpenCVObstacleAvoidanceNode::handleImage, this);
    timer_ = nh_.createTimer(ros::Duration(0.1), &OpenCVObstacleAvoidanceNode::handleTimer, this);
    ROS_INFO("opencv_obstacle_avoidance_cpp: image=%s cmd=%s debug=%s", image_topic_.c_str(), cmd_topic_.c_str(), publish_debug_ ? "true" : "false");
  }

 private:
  void loadParams() {
    private_nh_.param<std::string>("image_topic", image_topic_, "/usb_cam/image_raw");
    private_nh_.param<std::string>("cmd_vel_topic", cmd_topic_, "/cmd_vel");
    private_nh_.param<std::string>("debug_image_topic", debug_topic_, "debug");
    private_nh_.param("publish_debug_image", publish_debug_, false);
    private_nh_.param("forward_speed", forward_speed_, 0.08);
    private_nh_.param("slow_speed", slow_speed_, 0.04);
    private_nh_.param("turn_speed", turn_speed_, 0.45);
    private_nh_.param("side_steer_gain", side_steer_gain_, 0.35);
    private_nh_.param("stop_on_no_image", stop_on_no_image_, true);
    private_nh_.param("no_image_timeout", no_image_timeout_, 1.0);
    private_nh_.param("floor_sample_y0", floor_sample_y0_, 0.82);
    private_nh_.param("floor_sample_y1", floor_sample_y1_, 0.96);
    private_nh_.param("roi_y0", roi_y0_, 0.38);
    private_nh_.param("roi_y1", roi_y1_, 0.88);
    private_nh_.param("roi_x0", roi_x0_, 0.12);
    private_nh_.param("roi_x1", roi_x1_, 0.88);
    private_nh_.param("hsv_floor_dist_thresh", hsv_floor_dist_thresh_, 42.0);
    private_nh_.param("edge_thresh", edge_thresh_, 70);
    private_nh_.param("dark_v_thresh", dark_v_thresh_, 55);
    private_nh_.param("min_saturation", min_saturation_, 35);
    private_nh_.param("center_stop_ratio", center_stop_ratio_, 0.18);
    private_nh_.param("side_slow_ratio", side_slow_ratio_, 0.24);
    private_nh_.param("clear_hysteresis", clear_hysteresis_, 0.04);
    private_nh_.param("min_obstacle_area", min_obstacle_area_, 600);
  }

  void handleTimer(const ros::TimerEvent&) {
    if (!stop_on_no_image_ || last_image_time_.isZero()) {
      return;
    }
    const double age = (ros::Time::now() - last_image_time_).toSec();
    if (age > no_image_timeout_) {
      ROS_WARN_THROTTLE(2.0, "opencv_obstacle_avoidance_cpp: no image %.2fs, stop", age);
      twist_pub_.publish(geometry_msgs::Twist());
    }
  }

  void handleImage(const sensor_msgs::Image::ConstPtr& msg) {
    last_image_time_ = ros::Time::now();
    cv_bridge::CvImageConstPtr cv_ptr;
    try {
      cv_ptr = cv_bridge::toCvShare(msg, "bgr8");
    } catch (const cv_bridge::Exception& ex) {
      ROS_WARN("opencv_obstacle_avoidance cv_bridge: %s", ex.what());
      return;
    }

    cv::Mat bgr = cv_ptr->image.clone();
    cv::Rect roi_box = makeRoiBox(bgr.cols, bgr.rows);
    cv::Mat mask = obstacleMask(bgr);
    cv::Mat roi_mask = mask(roi_box).clone();
    cv::Mat cleaned = cleanMask(roi_mask);
    const auto ratios = sectorRatios(cleaned);
    const geometry_msgs::Twist twist = decide(ratios[0], ratios[1], ratios[2]);
    twist_pub_.publish(twist);

    ROS_INFO_THROTTLE(1.0, "opencv_obstacle_avoidance_cpp: left=%.2f center=%.2f right=%.2f v=%.2f wz=%.2f", ratios[0], ratios[1], ratios[2], twist.linear.x, twist.angular.z);
    if (publish_debug_) {
      publishDebug(msg->header, bgr, roi_box, cleaned);
    }
  }

  cv::Rect makeRoiBox(int width, int height) const {
    int x0 = static_cast<int>(clampValue(roi_x0_, 0.0, 1.0) * width);
    int x1 = static_cast<int>(clampValue(roi_x1_, 0.0, 1.0) * width);
    int y0 = static_cast<int>(clampValue(roi_y0_, 0.0, 1.0) * height);
    int y1 = static_cast<int>(clampValue(roi_y1_, 0.0, 1.0) * height);
    if (x1 <= x0) { x0 = 0; x1 = width; }
    if (y1 <= y0) { y0 = static_cast<int>(0.4 * height); y1 = height; }
    return cv::Rect(x0, y0, x1 - x0, y1 - y0);
  }

  cv::Scalar floorModel(const cv::Mat& hsv) const {
    const int y0 = static_cast<int>(clampValue(floor_sample_y0_, 0.0, 1.0) * hsv.rows);
    const int y1 = static_cast<int>(clampValue(floor_sample_y1_, 0.0, 1.0) * hsv.rows);
    const int x0 = static_cast<int>(0.25 * hsv.cols);
    const int x1 = static_cast<int>(0.75 * hsv.cols);
    cv::Scalar mean_value = cv::mean(hsv(cv::Rect(x0, y0, std::max(1, x1 - x0), std::max(1, y1 - y0))));
    return mean_value;
  }

  cv::Mat obstacleMask(const cv::Mat& bgr) const {
    cv::Mat hsv;
    cv::Mat gray;
    cv::cvtColor(bgr, hsv, cv::COLOR_BGR2HSV);
    cv::cvtColor(bgr, gray, cv::COLOR_BGR2GRAY);
    const cv::Scalar floor = floorModel(hsv);

    cv::Mat mask(bgr.rows, bgr.cols, CV_8UC1, cv::Scalar(0));
    cv::Mat edges;
    cv::Canny(gray, edges, edge_thresh_, edge_thresh_ * 2);
    for (int y = 0; y < hsv.rows; ++y) {
      for (int x = 0; x < hsv.cols; ++x) {
        const cv::Vec3b p = hsv.at<cv::Vec3b>(y, x);
        double dh = std::abs(static_cast<double>(p[0]) - floor[0]);
        dh = std::min(dh, 180.0 - dh) * 1.4;
        const double ds = std::abs(static_cast<double>(p[1]) - floor[1]) * 0.35;
        const double dv = std::abs(static_cast<double>(p[2]) - floor[2]) * 0.55;
        const bool non_floor = (dh + ds + dv) > hsv_floor_dist_thresh_;
        const bool dark = p[2] < dark_v_thresh_;
        const bool saturated = p[1] > min_saturation_;
        if (non_floor && (dark || saturated || edges.at<unsigned char>(y, x) > 0)) {
          mask.at<unsigned char>(y, x) = 255;
        }
      }
    }
    cv::morphologyEx(mask, mask, cv::MORPH_OPEN, cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(3, 3)));
    cv::morphologyEx(mask, mask, cv::MORPH_CLOSE, cv::getStructuringElement(cv::MORPH_ELLIPSE, cv::Size(7, 7)));
    return mask;
  }

  cv::Mat cleanMask(const cv::Mat& roi) const {
    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(roi, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
    cv::Mat cleaned = cv::Mat::zeros(roi.size(), CV_8UC1);
    for (const auto& contour : contours) {
      if (cv::contourArea(contour) >= min_obstacle_area_) {
        cv::drawContours(cleaned, std::vector<std::vector<cv::Point>>{contour}, -1, cv::Scalar(255), cv::FILLED);
      }
    }
    return cleaned;
  }

  std::array<double, 3> sectorRatios(const cv::Mat& cleaned) const {
    const int part_w = std::max(1, cleaned.cols / 3);
    std::array<double, 3> ratios {0.0, 0.0, 0.0};
    for (int index = 0; index < 3; ++index) {
      const int x0 = index * part_w;
      const int x1 = (index == 2) ? cleaned.cols : std::min(cleaned.cols, x0 + part_w);
      const cv::Mat part = cleaned(cv::Rect(x0, 0, std::max(1, x1 - x0), cleaned.rows));
      ratios[index] = static_cast<double>(cv::countNonZero(part)) / std::max(1, part.rows * part.cols);
    }
    return ratios;
  }

  geometry_msgs::Twist decide(double left, double center, double right) {
    geometry_msgs::Twist twist;
    double stop_threshold = center_stop_ratio_;
    if (last_blocked_) {
      stop_threshold = std::max(0.01, stop_threshold - clear_hysteresis_);
    }
    const bool blocked = center >= stop_threshold;
    if (blocked) {
      twist.angular.z = (left < right) ? turn_speed_ : -turn_speed_;
    } else if (left >= side_slow_ratio_ || right >= side_slow_ratio_) {
      twist.linear.x = slow_speed_;
      twist.angular.z = -clampValue((left - right) * side_steer_gain_, -turn_speed_, turn_speed_);
    } else {
      twist.linear.x = forward_speed_;
    }
    last_blocked_ = blocked;
    return twist;
  }

  void publishDebug(const std_msgs::Header& header, cv::Mat& bgr, const cv::Rect& roi_box, const cv::Mat& cleaned) {
    cv::rectangle(bgr, roi_box, cv::Scalar(0, 255, 255), 2);
    cv_bridge::CvImage out(header, "bgr8", bgr);
    debug_pub_.publish(out.toImageMsg());
  }

  ros::NodeHandle nh_;
  ros::NodeHandle private_nh_;
  ros::Publisher twist_pub_;
  ros::Publisher debug_pub_;
  ros::Subscriber image_sub_;
  ros::Timer timer_;
  ros::Time last_image_time_;
  std::string image_topic_;
  std::string cmd_topic_;
  std::string debug_topic_;
  bool publish_debug_ = false;
  double forward_speed_ = 0.08;
  double slow_speed_ = 0.04;
  double turn_speed_ = 0.45;
  double side_steer_gain_ = 0.35;
  bool stop_on_no_image_ = true;
  double no_image_timeout_ = 1.0;
  double floor_sample_y0_ = 0.82;
  double floor_sample_y1_ = 0.96;
  double roi_y0_ = 0.38;
  double roi_y1_ = 0.88;
  double roi_x0_ = 0.12;
  double roi_x1_ = 0.88;
  double hsv_floor_dist_thresh_ = 42.0;
  int edge_thresh_ = 70;
  int dark_v_thresh_ = 55;
  int min_saturation_ = 35;
  double center_stop_ratio_ = 0.18;
  double side_slow_ratio_ = 0.24;
  double clear_hysteresis_ = 0.04;
  int min_obstacle_area_ = 600;
  bool last_blocked_ = false;
};

}  // namespace simple_robotic_arm

int main(int argc, char** argv) {
  ros::init(argc, argv, "opencv_obstacle_avoidance");
  simple_robotic_arm::OpenCVObstacleAvoidanceNode node;
  ros::spin();
  return 0;
}
